src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509CertificateReader.cs

```diff
@@ -21,29 +21,54 @@ internal sealed class OpenSslX509CertificateReader : ICertificatePal
         private X500DistinguishedName _subjectName;
         private X500DistinguishedName _issuerName;
 
-        internal unsafe OpenSslX509CertificateReader(byte[] data)
+        internal OpenSslX509CertificateReader(SafeX509Handle handle)
         {
-            fixed (byte* pDataFixed = data)
+            // X509_check_purpose has the effect of populating the sha1_hash value,
+            // and other "initialize" type things.
+            bool init = Interop.libcrypto.X509_check_purpose(handle, -1, 0);
+
+            if (!init)
             {
-                byte* pData = pDataFixed;
-                byte** ppData = &pData;
+                throw new CryptographicException(Interop.libcrypto.GetOpenSslErrorString());
+            }
 
-                _cert = Interop.libcrypto.d2i_X509(IntPtr.Zero, ppData, data.Length);
+            _cert = handle;
+        }
+
+        internal unsafe OpenSslX509CertificateReader(byte[] data)
+        {
+            SafeX509Handle cert;
 
-                if (_cert.IsInvalid)
+            // If the first byte is a hyphen then this is likely PEM-encoded,
+            // otherwise it's DER-encoded (or not a certificate).
+            if (data[0] == '-')
+            {
+                using (SafeBioHandle bio = Interop.libcrypto.BIO_new(Interop.libcrypto.BIO_s_mem()))
                 {
-                    throw new CryptographicException();
+                    Interop.libcrypto.BIO_write(bio, data, data.Length);
+                    cert = Interop.libcrypto.PEM_read_bio_X509_AUX(bio, IntPtr.Zero, IntPtr.Zero, IntPtr.Zero);
                 }
+            }
+            else
+            {
+                cert = Interop.libcrypto.OpenSslD2I(Interop.libcrypto.d2i_X509, data);
+            }
 
-                // X509_check_purpose has the effect of populating the sha1_hash value,
-                // and other "initialize" type things.
-                bool init = Interop.libcrypto.X509_check_purpose(_cert, -1, 0);
+            if (cert.IsInvalid)
+            {
+                throw new CryptographicException();
```

Comment:
I was thinking about that this morning.  It'd probably want to be a bit more complicated:

```C#
internal static void ThrowOpenSslCryptographicException()
{
    int error = ERR_get_error();
    int last = error;

    while (last != 0)
    {
        last = ERR_get_error();
    }

    throw error == 0 ? new CryptographicException() : new CryptographicException(GetOpenSslErrorString(error));
}
```

in order to deal with any error codes set while we hadn't bothered checking.  But, I think we're going to need to do something complicated in OpenSsl for error management; I'm pretty sure that their stack is cross-threaded, and we'd presumably want it to be unique per execution context.  I'll open an issue about the threading problem, and a separate one to make the helper function and clean up existing usages.